# HW 4

## Linear Regression

Simple linear regression is a way to describe a relationship between two variables through an equation of a straight line, called line of best fit, that most closely models the relationship. The following linear regression formula provides the least-squares, best-fit equation for the line:

$y=a+bx$ where $\displaystyle b = \frac{\sum_{i=1}^{n}x_iy_i-n\bar{x}\bar{y}}{\sum_{i=1}^{n}x_i^2-n\bar{x}^2}$ and $a=\bar{y}-b\bar{x}$, and $\bar{x}$ and $\bar{y}$ are the means of the x and y coordinates.

In [30]:
import numpy as np

def point_slope_factory(_ps):
    """return a function which compute y = mx + b, where m is the slope and b the y-intercept of a line"""
    def _f(_x):
        return _ps[1] + _ps[0] * _x
    return _f

def mean(_pts):
    """mean of n points (_pts is a 2-dimensional numpy.ndarray)"""
    _n = _pts.shape[0]  # the number of points
    _m = _pts[0].shape[0]  # the dimension of a point
    _u = np.zeros((_m,), np.float64)  # initialize u to zeros
    for _k in range(_n):
        _u += _pts[_k]
    _u /=_n  # divide by n
    return _u

def regression_line(_pts):
    _n = _pts.shape[0]  # the number of points
    _u = mean(_pts)  # the mean of each dimension
    _x = _pts.T[0]  # all x-coordinates (T property is transpose)
    _y = _pts.T[1]  # all y-coordinates
    _nm = sum(_x * _y) - _n * _u[0] * _u[1]  # numerator
    _dm = sum(_x * _x) - _n * _u[0] * _u[0]  # denominator
    _b = _nm / _dm
    _a = _u[1] - _b * _u[0]
    return [_b, _a]  # slope and y_intercept of the least-squares regression line

def covariance(_pts):
    """covariance of n points (_pts is a 2-dimensional numpy.ndarray) computed using the outer product algorithm"""
    _n = _pts.shape[0]  # n is the number of m-dimensional points
    _m = _pts[0].shape[0]  # the dimension of the points
    _mu = mean(_pts)  # m-dimensional means
    _cvm = np.zeros((_m, _m), np.float64)  # m by m matrix of zeros
    for _k in range(_n):
        _mm = _pts[_k] - _mu  # subtract the means from the point
        _cvm += np.outer(_mm, _mm)  # accumulate the outer product of mm with itself
    _cvm /= _n  # divide by n
    return _cvm;      

1) Compute the regression line slope and y-intercept.

In [31]:
    
# points
pts = np.array([[2,2], [0,0], [-2,-2], [-1,1], [1,-1]])
lp = regression_line(pts)
print('regression line slope={}, y-intercept={}'.format(lp[0], lp[1]))


regression line slope=0.6, y-intercept=0.0


2) Compute the least-squares error, which is simply the sum of the squares of the y-differences between each original point and the best-fit regression line.

In [32]:
rgl = point_slope_factory(lp)
x = pts.T[0]
y = pts.T[1]
ye = [(y[i] - rgl(x[i])) ** 2  for i in range(len(x))]
print('least squares error: {}'. format(sum(ye)))

least squares error: 6.400000000000001


3) Because some might think the best-fit regression line should have slope = 1 and go through the points (2,2) and (-2,-2), compute the least-squares error to this line to show that it is greater than the least- squares error obtained above.

In [33]:
sgl = point_slope_factory([1.0, 0.0])  # slope and y-intercept
ye2 = [(y[i] - sgl(x[i])) ** 2  for i in range(len(x))]
print('least squares error of 45-degree line: {}'. format(sum(ye2)))

least squares error of 45-degree line: 8.0


In [35]:
cvm = covariance(pts)
print('covariance matrix: {}'. format(cvm))

covariance matrix: [[2.  1.2]
 [1.2 2. ]]


Eigenvalues of the covariance matrix. $\det\begin{pmatrix}2-\lambda&1.2\\1.2&2-\lambda\end{pmatrix}=(2-\lambda)(2-\lambda)-1.2\cdot 1.2=\lambda^2-4\lambda+2.56$.  The roots of this equation are $(3.2,0.8)$.  The largest eigenvalue is $3.2$.  The corresponding eigenvector is $$\begin{pmatrix}2&1.2\\1.2&2\end{pmatrix}\begin{pmatrix}e_1\\e_2\end{pmatrix}=3.2\begin{pmatrix}e_1\\e_2\end{pmatrix}$$
$$\begin{pmatrix}2e_1+1.2e_2\\1.2e_1+2e_2\end{pmatrix}=\begin{pmatrix}3.2e_1\\3.2e_2\end{pmatrix}\text{  solution: }e_1=e_2$$

Therefore, the unit eigenvector is $(1,1)$. The PCA line is the line with the slope equal to the angle of the eigenvector, through the point $(\bar{x},\bar{y})$.